sdlib/d/gc/extent.d

```diff
@@ -116,14 +142,51 @@ private:
 		bits |= ulong(arenaIndex) << 32;
 
 		if (ec.isSlab()) {
-			import d.gc.bin;
-			bits |= ulong(binInfos[ec.sizeClass].slots) << 48;
+			bits |= slabSlots << 48;
 
+			// Clear all slab occupancy and any meta flags as well
 			slabData.clear();
+		} else {
+			clearLarge();
 		}
 	}
 
 public:
+	void clearLarge() {
+		meta.large.allocSize = 0;
+		meta.large.canAppend = false;
+		meta.large.finalizer = null;
+	}
+
+	@property
+	ref bool appendable() {
+		assert(!isSlab(), "appendable accessed on slab!");
+		return meta.large.canAppend;
+	}
+
+	@property
+	bool finalizable() {
+		return finalizer != null;
+	}
+
+	@property
+	ref ulong allocSize() {
+		assert(!isSlab(), "allocSize accessed on slab!");
+		return meta.large.allocSize;
+	}
+
+	@property
+	ref void* finalizer() {
+		assert(!isSlab(), "finalizer accessed on slab!");
+		return meta.large.finalizer;
+	}
+
+	@property
+	ulong slabSlots() const {
+		assert(isSlab(), "slabSlots accessed on non slab!");
+		return ulong(binInfos[sizeClass].slots);
+	}
```

Comment:
Every time this is called, you take the risk of taking one cache miss. There is no reason to make it look like member data, and more generally, I doubt it's actually what's needed because I doubt most of these computation ought to happen in this class. In fact, #283 kinda gives it away, as computation there is redundant with computation here, so even without the code actually doing anything, it's already suspicious.

Reply:
How can one determine which meta bits are valid for a given slab alloc size class without knowing the slot size? And it doesn't seem that there is any convenient place to cache it locally in `Extent`. And where would you have the computation happen if not in `Extent` ?

Reply:
Also, there is nothing redundant about this piece vs. #283. In the latter, we determine the effective size of a slab alloc required to support the flags shown here. Let's illustrate with excerpt from the size class dumper:

```
id: 0   size: 8          needPages: 1    slots: 512
id: 1   size: 16        needPages: 1    slots: 256
id: 2   size: 24        needPages: 3    slots: 512
id: 3   size: 32        needPages: 1    slots: 128
id: 4   size: 40        needPages: 5    slots: 512
id: 5   size: 48        needPages: 3    slots: 256
id: 6   size: 56        needPages: 7    slots: 512
id: 7   size: 64        needPages: 1    slots: 64
id: 8   size: 80        needPages: 5    slots: 256
id: 9   size: 96        needPages: 3    slots: 128
id: 10  size: 112       needPages: 7    slots: 256
```

With the given scheme, classes and 0, 2, 4, 6 leave no room for any flags.
1, 5, 8, and 10 allow only one flag per slab slot.
All others permit two flags per slot.

In #283 we compute any required up-sizing for a slab alloc to allow it to utilize the requested flags.
Whereas here we access the flags per se.

Reply:
Thinking about it, I can hardcode the classes in the predicates in `Extent`; this is probably what you meant when said "redundant". But this would leave less flexibility to modify the size classes later. Is this what we want?

Reply:
Which means you can do `(sc & 0x01) || (sc > 6)` to know which size are suitable for appendable allocations. you can use a unittest to ensure implementation match the intended behavior (256 entries or less).

Reply:
@deadalnix Excellent point, thank you!